### PR TITLE
chapter_14

### PR DESCRIPTION
**Pull request by chuan97sun**

### CODE REVIEW

**Comment by chuan97sun** [comment]
ready

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -50,6 +53,19 @@ def destroy
...
+    @users = @user.following.page params[:page]
+    render "show_follow"
+  end
+
+  def followers
+    @title = "Followers"
+    @user = User.find(params[:id])
+    @users = @user.followers.page params[:page]
+    render "show_follow"
+  end
```
Comment (chungpv-1008) [inline app/controllers/users_controller.rb]: Why don't you create follows_controller.rb, then put these in there?
```suggestion
  def following
    @title = "Following"
    @users = @user.following.page params[:page]
    render "show_follow"
  end

  def followers
    @title = "Followers"
    @users = @user.followers.page params[:page]
    render "show_follow"
  end
  
  private
  
  def find_user
     @user = User.find id: params[:id]
     return if @user.present?
     
     redirect...
  end
```

File: app/models/user.rb [hunk 2]
```diff
@@ -1,5 +1,10 @@
 class User < ApplicationRecord
   has_many :microposts, dependent: :destroy
+  has_many :active_relationships, class_name: "Relationship", foreign_key: "follower_id", dependent: :destroy
+  has_many :passive_relationships, class_name: "Relationship", foreign_key: "followed_id",dependent: :destroy
```
Comment (chungpv-1008) [inline app/models/user.rb]:
```suggestion
  has_many :active_relationships, class_name: Relationship.name, foreign_key: :follower_id, dependent: :destroy
  has_many :passive_relationships, class_name: Relationship.name, foreign_key: :followed_id,dependent: :destroy                                
```

File: app/models/user.rb [hunk 3]
```diff
@@ -1,5 +1,10 @@
 class User < ApplicationRecord
   has_many :microposts, dependent: :destroy
+  has_many :active_relationships, class_name: "Relationship", foreign_key: "follower_id", dependent: :destroy
+  has_many :passive_relationships, class_name: "Relationship", foreign_key: "followed_id",dependent: :destroy                                
+  has_many :following, through: :active_relationships, source: :followed
+  has_many :followers, through: :passive_relationships, source: :follower
```
Comment (chungpv-1008) [inline app/models/user.rb]:
```suggestion
  has_many :followers, through: :passive_relationships
```

File: app/views/relationships/create.js.erb [hunk 4]
```diff
@@ -0,0 +1,2 @@
+$('#follow_form').html('<%= j render "users/unfollow" %>');
```
Comment (chungpv-1008) [inline app/views/relationships/create.js.erb]:
```suggestion
$('#follow_form').html('<%= j render 'users/unfollow' %>');
```

File: app/views/relationships/destroy.js.erb [hunk 5]
```diff
@@ -0,0 +1,2 @@
+$('#follow_form').html('<%= j render "users/follow" %>');
```
Comment (chungpv-1008) [inline app/views/relationships/destroy.js.erb]: similar

File: app/views/users/_follow.html.erb [hunk 6]
```diff
@@ -0,0 +1,4 @@
+<%= form_for(current_user.active_relationships.build) do |f| %>
```
Comment (chungpv-1008) [inline app/views/users/_follow.html.erb]:
```suggestion
<%= form_for current_user.active_relationships.build do |f| %>
```

File: app/views/users/_follow_form.html.erb [hunk 7]
```diff
@@ -0,0 +1,9 @@
+<% unless current_user?(@user) %>
```
Comment (chungpv-1008) [inline app/views/users/_follow_form.html.erb]:
```suggestion
<% unless current_user? @user %>
```

File: app/views/users/_follow_form.html.erb [hunk 8]
```diff
@@ -0,0 +1,9 @@
...
+  <div id="follow_form">
+  <% if current_user.following?(@user) %>
```
Comment (chungpv-1008) [inline app/views/users/_follow_form.html.erb]:
```suggestion
  <% if current_user.following? @user %>
```

File: app/views/users/_unfollow.html.erb [hunk 9]
```diff
@@ -0,0 +1,3 @@
+<%= form_for(current_user.active_relationships.find_by(followed_id: @user.id), html: { method: :delete }) do |f| %>
```
Comment (chungpv-1008) [inline app/views/users/_unfollow.html.erb]:
```suggestion
<%= form_for current_user.active_relationships.find_by(followed_id: @user.id), html: { method: :delete } do |f| %>
```

**Comment by chuan97sun** [comment]
ready

**Review by chungpv-1008 — COMMENTED** [review]
Why can't I see 2 actions 'following' and 'followers'

File: app/models/relationship.rb [hunk 10]
```diff
@@ -0,0 +1,6 @@
+class Relationship < ApplicationRecord
+  belongs_to :follower, class_name: "User"
```
Comment (ngocvt-0484) [inline app/models/relationship.rb]: User.name

File: app/models/relationship.rb [hunk 11]
```diff
@@ -0,0 +1,6 @@
...
+  belongs_to :followed, class_name: "User"
```
Comment (ngocvt-0484) [inline app/models/relationship.rb]: User.name

**Comment by chuan97sun** [comment]
rready

File: app/assets/stylesheets/custom.scss [hunk 12]
```diff
@@ -122,6 +122,44 @@ aside {
...
+    }
+    &:hover {
+      text-decoration: none;
+      color: blue;
+    }
+  }
+  strong {
+    display: block;
+  }
+}
```
Comment (saobv-2223) [inline app/assets/stylesheets/custom.scss]: image

File: app/controllers/follows_controller.rb [hunk 13]
```diff
@@ -0,0 +1,25 @@
+class FollowsController < ApplicationController
+
```
Comment (saobv-2223) [inline app/controllers/follows_controller.rb]:
```suggestion
```

File: app/controllers/follows_controller.rb [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+   private
+  
+  def find_user
+    @user = User.find id: params[:id]
+    return if @user.present?
+     
+
+    flash[:danger] = "not the user"
+    redirect_to root_path
```
Comment (saobv-2223) [inline app/controllers/follows_controller.rb]:
```suggestion
    @user = User.find id: params[:id]
    return if @user.present?
     
    flash[:danger] = "not the user"
    redirect_to root_path
```

File: app/controllers/users_controller.rb [hunk 15]
```diff
@@ -14,6 +14,9 @@ def new
   
   def show
     @microposts = @user.microposts.recent_posts.page params[:page]
+    @follow = current_user.active_relationships.build
+    @unfollow = current_user.following
+
```
Comment (saobv-2223) [inline app/controllers/users_controller.rb]:
```suggestion
```

**Comment by chuan97sun** [comment]
ready

File: config/routes.rb [hunk 16]
```diff
@@ -10,5 +10,11 @@
     resources :account_activations, only: :edit
     resources :password_resets, except: %i(index show destroy)
     resources :microposts, only: %i(create destroy)
+    resources :relationships, only: %i(create destroy)
+    resources :users do
+      member do
+        get :following, :followers
```
Comment (chungpv-1008) [inline config/routes.rb]:
```suggestion
        get :following, to: "follows#following"
        get :followers, to: "follows#followers"
```

**Pull request merged**